lightning/src/ln/peer_handler.rs

```diff
@@ -1088,15 +1095,107 @@ impl<Descriptor: SocketDescriptor> PeerManager<Descriptor> {
 			}
 		};
 	}
+
+// return a set of descriptors for all the peers in our peer manager
+fn get_des(&self) -> Vec<Descriptor> {
+	let mut des_set: Vec<Descriptor> = Vec::new();
+	let mut peers_lock = self.peers.lock().unwrap();
+	let peers = peers_lock.borrow_parts();
+			
+	for (Descriptor, mut Peer) in peers.peers.iter_mut() {
+
+				//read all events that have been sent from the Peer
+				let mut descriptor = Descriptor.clone();
+				des_set.push(descriptor);
+	}
+	des_set
+
+}
+
+//take a vector of descriptors and performs do_read_event on each of them
+fn mass_do_read_event(&self, mut des_set: Vec<Descriptor>)
+{
+	for (mut Descriptor) in des_set.iter_mut(){
+		let data: Vec<u8> = Vec::new();
+				
+		let res = match self.do_read_event(&mut Descriptor, data){
+					Ok(pause_read) => pause_read,
+					Err(e) => panic!("something is wrong"),
+		};
+	}
+
+}
+
+// iterate through our peers, if there are outstanding pings add a descriptor for the peer to a vector to be removed
+//otherwise ping all peers
+fn handle_pings(&self)-> Vec<Descriptor>{
+
+			let mut des_set: Vec<Descriptor> = Vec::new();
+			let mut peers_lock = self.peers.lock().unwrap();
+			let peers = peers_lock.borrow_parts();
+			
+			for (Descriptor, mut Peer) in peers.peers.iter_mut() {
+
+				// Disconect the Peer if there is an outstanding ping for which we have not been ponged
+				if Peer.ping_tracker > 0 {
+					let mut descriptor = Descriptor.clone();
```

Comment:
So, if i understood you correctly we would be looking for something along the lines of:

fn disconnect_event_internal_v2<'a> (&self, peers: MutPeerHolder<'a, Descriptor> , descriptor: &Descriptor, no_connection_possible: bool)

I am assuming the reason we are doing it that way is so we do not require the lock as we are passing in the MutPeerHolder, I tried to do that and I ran into some problems, it seems If i try to use disconnect_event_internal_v2 from inside the loop I can't because the iterator already takes ownership of the MutPeerHolder... I tried to just copy the entire MutPeerHolder and use that but the structure does not implement the copy trait.... 

of course I could implement the copy trait but it would be similar to just making my own set of descriptors?,

Reply:
Ah, then just copy the body of disconnect_event_internal. Its like a few lines of code, as long as you comment in both places that the code is duplicated elsewhere the DRY-ness of it isn't the most critical.

Reply:
sounds good